### DISCUSSION TITLE
Undo / redo for interactive sessions

### DISCUSSION THREAD

**Discussion by k4pran** [ManimCommunity/manim#2797] [Suggestions and Proposals]
## Description of proposed feature

When using interactive embed with the opengl renderer it would be good to have a history of states that you can undo / redo. As you build a scene one step at a time it will save you needing to restart the scene again if you want to go back and make changes.

## How can the new feature be used?

It makes interactive sessions more interactive and flexible, allowing you to try different things in the scene and undoing it to the previous state.

**Comment by TRoboto** [comment]
You can achieve something similar with the `self.skip_animation_preview` variable in any scene, just change its value depending on what you want to play. Additionally, when making a change to a scene, leave the window open as it will automatically update when the file is changed.

**Comment by k4pran** [comment]
> You can achieve something similar with the `self.skip_animation_preview` variable in any scene, just change its value depending on what you want to play. Additionally, when making a change to a scene, leave the window open as it will automatically update when the file is changed.

I'm not sure how to achieve it with `self.skip_animation_preview`? If I set that to True the scene doesn't appear? I also don't think it is the same when using the file observer for a few reasons, the first is that you may not want to alter the source code. Maybe you are just using some source code as a base, then interactively changing the scene, but you may want to leave the source as is. 

The second is that changing the file causes the scene to rerun, I think it would be more interactive to do it in place, just like adding or removing an mobject. Another benefit is that using the commandline is more flexible than changing the source code.
